### ISSUE TITLE
Likelihood in VAELoss 

### ISSUE THREAD

**Issue by shan-huang** [noveens/svae_cf#1]
In cell 10 of your notebook, when calculating likelihood in VAELoss, you calculated `num_ones = float(torch.sum(y_true_s[0, 0]))` and used it as the denominator of the sum of negative log likelihood loss. Since the shape of `y_true_s` is `[1, seq_len, total_items]`, I think `num_ones` only counts the ones in the first timestep? 

Would you explain a bit more please? 
1. Why need to count the ones and use it as a denominator? I assume it's sort of a normalization?
2. Why only count the number of ones in the first time step, instead of all time steps? (Maybe I misunderstood the dimension?)

Thank you very much.

**Comment by noveens** [comment]
Hi @shan-huang,

First of all, sincere apologies for the (very) late reply, I didn't seem to get an email when you posted this issue.
This is a good question, and my reason for including `num_ones` in the denominator is as follows:
- For the `next_k` loss type, `num_ones` is just counting `k`, and I want to obtain the average loss per next-k items rather than the sum, hence the normalization.

However, if we were to use the `postfix` version of the loss, this normalization would definitely be incorrect and we would need to compute the sums for all timesteps: `num_ones = float(torch.sum(y_true_s[0], dim = -1))` and divide by that. But since the paper anyways suggests using the `next_k` loss due to better performance, I guess that's the reason no one pointed this bug out before.

Anyways, much thanks for bringing this bug to my attention and I will fix it if someone else raises this issue as well.

**Comment by shan-huang** [comment]
Thank you for your explanation, @noveens ! I still can't quite get my head around this, even with the case in `next_k`.

When you calculated the loss, the numerator 
```
torch.sum(
            -1.0 * y_true_s.view(dec_shape[0] * dec_shape[1], -1) * \
            decoder_output.view(dec_shape[0] * dec_shape[1], -1)
        )
```
sums up the losses of `next_k` for all time steps, or all `seq_len`. And the denominator is a constant `k`. In that case, users who viewed more movies will have a larger loss, since they have bigger `seq_len`. Is that expected?

Thanks again!

**Comment by noveens** [comment]
A good catch again, @shan-huang. Yes, this is by design. I remember experimenting with both the mean of all timesteps and the sum of all timesteps and noticed a slightly better performance with the sum of losses across `seq_len`. Not sure if this would be applicable to all datasets though and I suspect the mean to work better for datasets where the item frequency distribution has a very big tail. But, it's always safe to try both the options out.

Hope this helps!

**Comment by shan-huang** [comment]
It's very helpful, thanks for your confirmation, @noveens ! Take care.
